# Notebook 04 — Machine Learning

Mục tiêu:
1. So sánh 4 mô hình (Dummy baseline + Linear + RandomForest + GradientBoosting).
2. Đánh giá bằng 5-fold CV trên train + 1 lần đánh giá trên test.
3. Phân tích 10 trường hợp sai lớn nhất.
4. K-Means chọn K theo silhouette.
5. Demo RecommendationEngine với 3 hồ sơ nhu cầu.

In [1]:
import json
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split

from src.features import build_preprocessor, get_feature_names
from src.predictor import PricePredictor, cv_metrics
from src.segmenter import KMeansSegmenter, pick_k_by_silhouette
from src.recommender import RecommendationEngine

RANDOM_STATE = 42
FIG = Path('reports/figures')
FIG.mkdir(parents=True, exist_ok=True)

df = pd.read_csv('data/processed/listings_with_amenities.csv').dropna(subset=['price_per_m2']).copy()
print('Loaded:', df.shape)

Loaded: (1773, 41)


## 1. Train / Test split

In [2]:
NUMERIC = ['area_m2', 'bedrooms', 'bathrooms', 'direction_code', 'balcony_code',
           'furnishing_code', 'legal_code', 'image_count', 'apartment_type']
CATEGORICAL = ['district_clean', 'direction_clean', 'project_name']
FEATURES = NUMERIC + CATEGORICAL

df[FEATURES] = df[FEATURES].fillna({c: 'missing' for c in CATEGORICAL})
X = df[FEATURES]
y = df['price_per_m2'].astype(float)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=RANDOM_STATE)
print('Train:', X_train.shape, 'Test:', X_test.shape)

pre = build_preprocessor(NUMERIC, CATEGORICAL)
pre.fit(X_train)
X_train_t = pre.transform(X_train)
X_test_t = pre.transform(X_test)
print('Features sau transform:', X_train_t.shape[1])

Train: (1418, 12) Test: (355, 12)
Features sau transform: 47


## 2. So sánh 4 mô hình — 5-fold CV + Test

In [3]:
rows = []
for name in ['dummy', 'linear', 'rf', 'gbr']:
    m = cv_metrics(X_train_t, y_train.values, model_name=name, log_target=True,
                   n_splits=5, random_state=RANDOM_STATE)
    pred = PricePredictor(model_name=name, log_target=True, random_state=RANDOM_STATE).fit(
        X_train_t, y_train.values
    )
    test_m = pred.evaluate(X_test_t, y_test.values)
    rows.append({
        'model': name,
        'CV_MAE_mean': m['mae_mean'],
        'CV_MAE_std': m['mae_std'],
        'CV_RMSE_mean': m['rmse_mean'],
        'CV_R2_mean': m['r2_mean'],
        'Test_MAE': test_m['mae'],
        'Test_RMSE': test_m['rmse'],
        'Test_R2': test_m['r2'],
    })

results = pd.DataFrame(rows)
results.round(3)

,model,CV_MAE_mean,CV_MAE_std,CV_RMSE_mean,CV_R2_mean,Test_MAE,Test_RMSE,Test_R2
0,dummy,1.884115e+07,1469898.974,3.408969e+07,-0.048,1.851669e+07,3.713354e+07,-0.041
1,linear,1.407244e+07,1049522.469,2.812106e+07,0.294,1.491922e+07,3.368822e+07,0.143
2,rf,1.299226e+07,724428.666,2.680897e+07,0.357,1.388349e+07,3.317713e+07,0.169
3,gbr,1.338326e+07,1162355.116,2.670725e+07,0.366,1.439704e+07,3.331663e+07,0.162


**Nhận xét (model comparison):**

**Kết quả trên test (355 mẫu):**
- **Random Forest** đạt Test R² = 0.169 (cao nhất) và Test MAE = 13.88tr/m² — nhỉnh hơn GBR một chút.
- **Gradient Boosting** đạt Test R² = 0.162, Test MAE = 14.40tr/m² — gần ngang RF.
- **Linear Regression** Test R² = 0.143, MAE = 14.92tr/m² — yếu hơn nhưng vẫn ổn.
- **Dummy baseline** Test R² = -0.041, MAE = 18.52tr/m² — baseline (predict median). Cả 3 model đều vượt rõ rệt.

**Hiện tượng overfitting nhẹ (CV R² > Test R²):**
- RF: CV R² = 0.357 → Test R² = 0.169 (giảm 0.188)
- GBR: CV R² = 0.366 → Test R² = 0.162 (giảm 0.204, giảm nhiều hơn RF)
- Linear: CV R² = 0.294 → Test R² = 0.143 (giảm 0.151)
- → GBR có dấu hiệu overfit hơi nhiều hơn RF; vì vậy RF **ổn định hơn** trên data mới.

**Ý nghĩa thực tế:**
- Sai số ~13-15 triệu/m² trên median ~53 triệu/m² = **khoảng 25%**.
- Mức R² ~0.17 là **khiêm tốn nhưng hợp lý** với data căn hộ — giá phụ thuộc nhiều yếu tố phi số (view, tầng, nội thất chi tiết, tiến độ bàn giao) mà model không có.
- **Kết luận**: RF và GBR tương đương; RF được chọn vì generalize nhỉnh hơn (gap CV-Test nhỏ hơn).

## 3. Phân tích 10 trường hợp sai lớn nhất (Random Forest)

In [4]:
best = PricePredictor(model_name='rf', log_target=True, random_state=RANDOM_STATE).fit(
    X_train_t, y_train.values
)
y_pred = best.predict(X_test_t)
errors_df = X_test.copy()
errors_df['actual_price_per_m2'] = y_test.values
errors_df['predicted_price_per_m2'] = y_pred
errors_df['abs_error'] = np.abs(errors_df['actual_price_per_m2'] - errors_df['predicted_price_per_m2'])
errors_df['pct_error'] = 100 * errors_df['abs_error'] / errors_df['actual_price_per_m2']
errors_df['listing_id'] = df.loc[y_test.index, 'listing_id'].values
errors_df['title'] = df.loc[y_test.index, 'title'].values
top10 = errors_df.nlargest(10, 'abs_error')[['listing_id', 'district_clean', 'area_m2',
                                             'bedrooms', 'actual_price_per_m2',
                                             'predicted_price_per_m2', 'pct_error', 'title']]
top10

,listing_id,district_clean,area_m2,bedrooms,actual_price_per_m2,predicted_price_per_m2,pct_error,title
1208,177641859,Quận 12,26.0,1.0,4.807692e+08,4.223649e+07,91.214810,Bán căn hộ dịch vụ có thang máy tại Quận 12 - ...
162,177622744,Quận Bình Thạnh,60.0,3.0,2.583333e+08,5.745550e+07,77.759160,bán nhà chung cư
73,177494566,Quận Bình Thạnh,122.4,3.0,2.287582e+08,7.240064e+07,68.350579,BÁN CĂN HỘ LANDMARK 81 - TẦNG 27 - SỞ HỮU LÂU ...
680,175165640,Thành phố Thủ Đức,64.0,1.0,2.156250e+08,9.846605e+07,54.334583,"Cần bán CHCC 1PN Empire City Thủ Thiêm, 13,8 t..."
339,177757531,Thành phố Thủ Đức,74.0,2.0,1.621622e+08,5.892784e+07,63.661163,Lumière Riverside 2PN FULL NT View Đẹp – Chỉ 1...
1692,177715817,Quận 7,82.0,2.0,1.463415e+08,5.482818e+07,62.534080,"Bán, cho thuê 2pn Midtown tầng cao suất SPA 높은..."
713,177850739,Thành phố Thủ Đức,140.0,3.0,5.142857e+07,1.272547e+08,147.439622,Cần Bán gấp căn hộ Duplex thông Tầng Hàng hiếm
109,177770338,Quận Bình Thạnh,65.0,2.0,1.215385e+08,5.585090e+07,54.046725,Bán căn hộ Landmark 3 Vinhomes Central Park vi...
1094,177866544,Thành phố Thủ Đức,51.0,1.0,1.176471e+08,5.243211e+07,55.432709,"Căn hộ Masteri An Phú 1PN hiếm gặp, giá sập sà..."
526,177766301,Thành phố Thủ Đức,71.0,2.0,1.338028e+08,7.023835e+07,47.506072,Bán gấp căn góc 2PN 71m2 tháp A6 Eaton Park Qu...


**Nhận xét (10 trường hợp sai):**

1. Phần lớn sai ở các căn **giá cực cao (~150–480 tr/m²)** — mô hình dự đoán thấp hơn nhiều (vì train data nghèo các ví dụ giá cực cao).
2. Các căn **diện tích lớn ở vị trí đắt đỏ** (Thủ Đức, Bình Thạnh) cũng sai nhiều — mô hình dựa vào feature "diện tích" thường đi kèm giá trung bình.
3. Các căn **thiếu nhiều thông tin** (direction, furnishing) → imputation điền median → có thể làm sai lệch.

**Hạn chế:** Dataset có **1773 dòng** (sau dropna `price_per_m2`) từ 1 nguồn (chotot) và 6 quận → mô hình thiếu thông tin về giá cực trị và khu vực rộng hơn. Cần mở rộng phạm vi để cải thiện.

## 4. K-Means — chọn K theo silhouette

In [5]:
best_k, scores = pick_k_by_silhouette(X_train_t, k_range=(3, 6), random_state=RANDOM_STATE)
print(f'Best K = {best_k}')
print(f'Scores: {scores}')

import matplotlib.pyplot as plt
fig, ax = plt.subplots(figsize=(8, 4))
ax.plot(list(scores.keys()), list(scores.values()), marker='o', color='darkgreen')
ax.set_title('Silhouette score theo K')
ax.set_xlabel('K')
ax.set_ylabel('Silhouette score')
ax.axvline(best_k, linestyle='--', color='red', label=f'best K={best_k}')
ax.legend()
plt.tight_layout()
plt.savefig(FIG / 'fig11_silhouette.png', dpi=120)
plt.close()
print('Saved fig11')

Best K = 4
Scores: {3: 0.07872245039347146, 4: 0.08297763779328028, 5: 0.0271645114275248, 6: -0.06124824181245423}


Saved fig11


In [6]:
seg = KMeansSegmenter(n_clusters=best_k, random_state=RANDOM_STATE).fit(X_train_t)
train_clusters = seg.predict(X_train_t)
test_clusters = seg.predict(X_test_t)

df_train = df.loc[X_train.index].copy()
df_train['cluster'] = train_clusters
df_test = df.loc[X_test.index].copy()
df_test['cluster'] = test_clusters
df_all_with_cluster = pd.concat([df_train, df_test], axis=0)
print('Cluster counts:')
print(df_all_with_cluster['cluster'].value_counts().sort_index())

Cluster counts:
cluster
0     165
1    1396
2      47
3     165
Name: count, dtype: int64


## 5. Mô tả cụm

In [7]:
all_X_t = pre.transform(df_all_with_cluster[FEATURES].fillna({c: 'missing' for c in CATEGORICAL}))
feat_names = get_feature_names(pre)
desc = seg.describe_segments(all_X_t, feat_names)
print('Cluster centers (mean feature value):')
desc.round(2)

Cluster centers (mean feature value):


,area_m2,bedrooms,bathrooms,direction_code,balcony_code,furnishing_code,legal_code,image_count,apartment_type,district_clean_Quận 12,...,project_name_Sunrise CityView,project_name_Sunshine Sky City,project_name_The Beverly Solari - Vinhomes Grand Park,project_name_The Era Town,project_name_The Manor - TP.Hồ Chí Minh,project_name_The Origami – Vinhomes Grand Park,project_name_Vinhomes Central Park,project_name_Vinhomes Grand Park,project_name_missing,project_name_infrequent_sklearn
__cluster__,,,,,,,,,,,,,,,,,,,,,
0,-0.12,-0.13,-0.14,0.57,1.04,0.26,-0.20,0.34,-0.05,0.03,...,0.00,0.00,0.04,0.00,0.00,0.0,0.02,0.36,0.10,0.37
1,0.04,0.02,0.04,0.19,0.08,-0.04,0.02,-0.04,0.03,0.11,...,0.01,0.01,0.01,0.03,0.01,0.0,0.03,0.00,0.21,0.51
2,-0.45,-0.20,-0.23,-0.09,0.15,0.05,0.03,-0.25,-0.16,0.00,...,0.00,0.00,0.00,0.00,0.00,1.0,0.00,0.00,0.00,0.00
3,-0.00,0.05,-0.18,-2.27,-1.70,0.26,-0.15,0.16,0.08,0.07,...,0.00,0.01,0.01,0.01,0.01,0.0,0.02,0.02,0.22,0.47


**Nhận xét:** Có 3 cụm với silhouette thấp (~0.09) → các cụm tách biệt không rõ. Điều này dự kiến vì đặc trưng BĐS thay đổi liên tục, không rời rạc. Tuy vậy, K-Means vẫn hữu ích để gợi ý vì bonus cụm giúp phân nhóm giá.

## 6. Demo RecommendationEngine — 3 hồ sơ nhu cầu

In [8]:
eng = RecommendationEngine()
profiles = [
    {
        'name': 'Gia đình trẻ, 3 tỷ, Thủ Đức / Bình Thạnh',
        'budget_vnd': 3e9,
        'target_bedrooms': 2,
        'target_area_m2': 65.0,
        'preferred_districts': ['Thành phố Thủ Đức', 'Quận Bình Thạnh'],
        'preferred_cluster': 0,
    },
    {
        'name': 'Nhà đầu tư, 5 tỷ, Quận 7 / Bình Tân',
        'budget_vnd': 5e9,
        'target_bedrooms': 2,
        'target_area_m2': 70.0,
        'preferred_districts': ['Quận 7', 'Quận Bình Tân'],
        'preferred_cluster': 1,
    },
    {
        'name': 'Người mua cao cấp, 7 tỷ, Thủ Đức / Quận 7',
        'budget_vnd': 7e9,
        'target_bedrooms': 3,
        'target_area_m2': 85.0,
        'preferred_districts': ['Thành phố Thủ Đức', 'Quận 7'],
        'preferred_cluster': 1,
    },
]

cluster_counts = df_all_with_cluster['cluster'].value_counts().to_dict()
for prof in profiles:
    pc = prof['preferred_cluster']
    if pc is not None and cluster_counts.get(pc, 0) < 100:
        print(f'⚠️  Profile "{prof["name"]}" dùng cluster {pc} chỉ có {cluster_counts.get(pc, 0)} tin — gợi ý sẽ thiếu bonus cluster.')
    recs = eng.recommend(
        df_all_with_cluster,
        budget_vnd=prof['budget_vnd'],
        target_bedrooms=prof["target_bedrooms"],
        target_area_m2=prof["target_area_m2"],
        preferred_districts=prof["preferred_districts"],
        preferred_cluster=prof["preferred_cluster"],
        top_k=5,
    )
    sep = '=' * 70
    print(f"\n{sep}\n{prof['name']}\n{sep}")
    display_cols = ['listing_id', 'district_clean', 'ward', 'total_price', 'area_m2', 'bedrooms', 'price_per_m2', 'amenity_score', 'cluster', 'score_total']
    print(recs[display_cols].to_string(index=False))


Gia đình trẻ, 3 tỷ, Thủ Đức / Bình Thạnh
 listing_id    district_clean                             ward  total_price  area_m2  bedrooms  price_per_m2  amenity_score  cluster  score_total
  177591150 Thành phố Thủ Đức Phường Long Thạnh Mỹ (Quận 9 cũ)   3050000000     64.0       2.0  4.765625e+07          403.4        0     2.452063
  177838754 Thành phố Thủ Đức Phường Long Thạnh Mỹ (Quận 9 cũ)   3190000000     65.0       2.0  4.907692e+07          403.4        0     2.436667
  175863420 Thành phố Thủ Đức      Phường Hiệp Phú (Quận 9 cũ)   2900000000     66.0       2.0  4.393939e+07          403.4        0     2.436636
  177920244 Thành phố Thủ Đức Phường Long Thạnh Mỹ (Quận 9 cũ)   2800000000     61.5       2.0  4.552846e+07          403.4        0     2.432604
  174651042 Thành phố Thủ Đức Phường Long Thạnh Mỹ (Quận 9 cũ)   2700000000     59.0       2.0  4.576271e+07          403.4        0     2.399218

Nhà đầu tư, 5 tỷ, Quận 7 / Bình Tân
 listing_id district_clean             ward  

**Nhận xét:** Hệ gợi ý trả về top-K theo `score_total` (gồm 4 thành phần: gần giá mục tiêu, gần diện tích mục tiêu, cùng cụm, tiện ích cao). Cả 3 profile đều có kết quả hợp lệ — phù hợp với phạm vi 6 quận căn hộ.

**Lưu ý quan trọng về `preferred_cluster`:** chỉ là **bonus điểm (+0.3)**, không phải hard filter. Nếu cluster quá nhỏ (ví dụ cluster 2 chỉ 47 tin), hệ vẫn recommend các tin thuộc cluster khác — bonus cluster sẽ không có. Profile cao cấp 7 tỷ ban đầu dùng cluster 2 → 0 tin khớp filter cứng → đã đổi sang cluster 1 (1396 tin).

**Lưu ý về district:** profile chỉ khuyến nghị được nếu `preferred_districts` chứa tên ĐÚNG với data (6 quận: Thủ Đức, Q7, Bình Thạnh, Bình Tân, Q12, Gò Vấp). Sai tên → recommend trả về rỗng.

## 7. Kết luận

- **Mô hình tốt nhất**: Random Forest với Test R² = 0.169, Test MAE = 13.88tr/m² — khiêm tốn nhưng vượt baseline rõ rệt (Dummy R² = -0.041). GBR ngang bằng (R² = 0.162, MAE = 14.40tr/m²) nhưng gap CV-Test lớn hơn → RF generalize nhỉnh hơn, **chọn RF**.
- **Hạn chế**: MAE còn cao (~13-15tr/m², ~25% median), sai nhiều ở giá cực trị và tin thiếu thông tin. Hiện tượng overfit nhẹ (CV R² = 0.357 vs Test R² = 0.169).
- **Phân cụm**: K được chọn tự động theo silhouette (K=4). Các cụm tách biệt không rõ (silhouette ~0.08) → dùng bonus nhẹ trong gợi ý. Cluster 2 chỉ 47 tin quá nhỏ → có thể gây recommend không khớp cluster.
- **Hệ gợi ý**: hoạt động đúng cho cả 3 profile (3 tỷ Thủ Đức/Bình Thạnh, 5 tỷ Q7/Bình Tân, 7 tỷ Thủ Đức/Q7) sau khi dùng district names thực tế trong data.
- **Hướng phát triển**: thu thập thêm dữ liệu (mở rộng 24 quận), thêm feature khoảng cách đến trung tâm, view, tầng, dùng mô hình XGBoost/LightGBM, regularization chống overfit.